### ISSUE TITLE
Error: statically known to be contained within the environment are missing

### ISSUE THREAD

**Issue by visox** [zio/zio#7506]
```
lazy val zioV = "2.0.3"
lazy val sttpV = "3.8.3"
...

libraryDependencies ++= Seq(
      "dev.zio" %% "zio" % zioV,
      ...
      "com.softwaremill.sttp.client3" %% "core" % sttpV,
      "com.softwaremill.sttp.client3" %% "zio" % sttpV,
)
```

Gives me an error on startup

```
timestamp=2022-11-07T12:41:30.424961Z level=ERROR thread=#zio-fiber-0 message="" cause="Exception in thread "zio-fiber-6,4" java.lang.Error: Defect in zio.ZEnvironment: Set(SttpBackend[=λ %A → ZIO[-Any,+Throwable,+A],+{package::WebSockets & ZioStreams}]) statically known to be contained within the environment are missing
	at zio.ZEnvironment.prune(ZEnvironment.scala:71)
	at zio.ZEnvironment.union(ZEnvironment.scala:95)
	at zio.ZLayer.$plus$plus$$anonfun$1(ZLayer.scala:66)
	at zio.ZIO.zipWith$$anonfun$1$$anonfun$1(ZIO.scala:2445)
	at zio.ZIO.map$$anonfun$1(ZIO.scala:941)
	at com.cognism.data.search.Main.run(Main.scala:11)
	at com.cognism.data.search.Main.run(Main.scala:11)
	at com.cognism.data.search.Main.run(Main.scala:11)
	at com.cognism.data.search.Main.run(Main.scala:11)
	at com.cognism.data.search.Main.run(Main.scala:11)
	at com.cognism.data.search.Main.run(Main.scala:11)
	at com.cognism.data.search.Main.run(Main.scala:11)
	at com.cognism.data.search.Main.run(Main.scala:11)"
```

guess i try lower versions

edit: lowering the zio version did help

**Comment by chessman** [comment]
I've got the same problem after upgrading from 2.0.2 to 2.0.3.

**Comment by narma** [comment]
It happens with Scala 3? In that case I have similar issue https://github.com/zio/zio/issues/7489 (actually not fixed yet)

**Comment by visox** [comment]
> It happens with Scala 3? In that case I have similar issue #7489 (actually not fixed yet)

not sure if relevant but yes i am working with scala 3

**Comment by adamgfraser** [comment]
Can you try on the latest snapshot?

**Comment by visox** [comment]
> Can you try on the latest snapshot?

is it publish anywhere or you need me to publish local ?

**Comment by adamgfraser** [comment]
There is a badge in the read me with a link to the latest snapshot.

**Comment by visox** [comment]
`lazy val zioV = "2.0.3+9-fd58759c-SNAPSHOT"`

sadly didnt help, the outcome is same as with `2.0.3`

**Comment by kyri-petrou** [comment]
From what I've gathered so far, the issue seems to be specifically with ZLayers and Scala 3. If a library was built with zio versions `<= 2.0.2` any ZLayers methods / vals they provide won't work in 2.0.3 / 2.0.4. A workaround is to not use the provided ZLayer but reconstruct it.

An example with sttp 3.8.3 (built with zio 2.0.2):

```scala
import sttp.client3.httpclient.zio.HttpClientZioBackend

val brokenLayer = HttpClientZioBackend.layer() // Will fail during runtime if used
val okLayer = ZLayer.scoped(HttpClientZioBackend.scoped()) // This is exactly what the `.layer()` method does but works
```

**Comment by ghostdogpr** [comment]
Had that issue as well when upgrading zio-query built with 2.0.3 in Caliban that's still using zio 2.0.2. Upgrading zio to 2.0.3 solves it but I can't because of https://github.com/zio/zio/issues/7496 😅

**Comment by bertlebee** [comment]
I managed this in 2.0.4, found this issue, easy fix right? Nope, went back to 2.0.2 and got the same thing
```<FAIL> Die(java.lang.Error: Defect in zio.ZEnvironment: Set(RpcReceiver[=LspClientMethod]) statically known to be contained within the environment are missing```

**Comment by marcinzh** [comment]
I got the same error, when upgrading from 2.0.4 to 2.0.5. Also Scala 3. ZLayer built with `fromZIO`. The module pattern I used is a bit atypical though.

**Comment by soujiro32167** [comment]
Got slapped by this with scala 2.13.10 and zio 2.0.5. Doing some...unusual layer composition, but since it compiles, it should work ;)

**Comment by Override-6** [comment]
Had the same issue today related with `ZioJdbcContext[+SqlIdiom,+NamingStrategy]`, i tried to downgrade from zio 2.0.6 until 2.0.0 but it didn't fixed the error, also tried to downgrade quill-zio-jdbc. My solution was to create a simple wrapper for this specific context : 

```scala
case class DatabaseContext(ctx: ZioJdbcContext[SqlIdiom, NamingStrategy]):
  export ctx.*
```

**Comment by neko-kai** [comment]
@Override-6 
This (for cases that are the same as https://github.com/zio/zio/issues/7489, caused by trailing `$` disappearing from object names in Tags) will keep happening until libraries update to izumi-reflect 2.2.0+ (Better 2.2.5+), so ping the maintainers to make new releases. I think at this point the Scala 3 version of izumi-reflect is stable-ish, so there will be no breakages like this in the future.

**Comment by jxnu-liguobin** [comment]
hello @neko-kai , Is there any new progress on this issue?

**Comment by neko-kai** [comment]
@jxnu-liguobin 
As above, it's fixed since izumi-reflect 2.2.0 (latest 2.3.2) and the corresponding zio version (latest 2.0.12). If the error happens in library code - it should be recompiled with latest zio version. If it happens in your code, you should update to latest zio version.
